### PR TITLE
[LoRA] add LoRA support to Lumina2 and fine-tuning script

### PR DESCRIPTION
**Pull request by sayakpaul**
# What does this PR do?

~Training script will follow in an immediate PR.~ Couldn't keep calm and ended up adding it.

TODOs

- [x] Training README
- [x] Training tests (can do in a follow-up PR, too)

Will update a successful wandb run page once available. But I think this is already ready to review now.

WandB: https://wandb.ai/sayakpaul/dreambooth-lumina2-lora/runs/gpyi2qsl?

### CODE REVIEW

**Comment by sayakpaul** [comment]
@zhuole1025 could you help verify the training?

Command:

<details>
<summary>Command</summary>

```bash
export MODEL_NAME="Alpha-VLLM/Lumina-Image-2.0"
export INSTANCE_DIR="dog"
export OUTPUT_DIR="trained-lumina2-lora"

CUDA_VISIBLE_DEVICES=0 accelerate launch train_dreambooth_lora_lumina2.py \
  --pretrained_model_name_or_path=$MODEL_NAME  \
  --instance_data_dir=$INSTANCE_DIR \
  --output_dir=$OUTPUT_DIR \
  --mixed_precision="bf16" \
  --instance_prompt="a photo of sks dog" \
  --resolution=1024 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=4 \
  --use_8bit_adam \
  --learning_rate=1e-5 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=500 \
  --validation_prompt="A photo of sks dog in a bucket" \
  --validation_epochs=25 \
  --seed="0"
```

</details>

Download the "dog" dataset like so:

```bash
from huggingface_hub import snapshot_download

local_dir = "./dog"
snapshot_download(
    "diffusers/dog-example",
    local_dir=local_dir, repo_type="dataset",
    ignore_patterns=".gitattributes",
)
```

Major question is do we need SD3 like `weighting_scheme` ("logit_normal") or do we not need that and we could simply follow what Flux, SANA, etc. do (current state).

Thanks in advance.

**Comment by zhuole1025** [comment]
> @zhuole1025 could you help verify the training?
> 
> Command:
> 
> Command
> Download the "dog" dataset like so:
> 
> ```shell
> from huggingface_hub import snapshot_download
> 
> local_dir = "./dog"
> snapshot_download(
>     "diffusers/dog-example",
>     local_dir=local_dir, repo_type="dataset",
>     ignore_patterns=".gitattributes",
> )
> ```
> 
> Major question is do we need SD3 like `weighting_scheme` ("logit_normal") or do we not need that and we could simply follow what Flux, SANA, etc. do (current state).
> 
> Thanks in advance.

Lumina is completely compatible with Flux, i.e., we use logit normal for training and dynamic shift for inference~

**Comment by sayakpaul** [comment]
> However, Lumina task time=1.0 as clean data, which is reserve to flux and sd3.

Not sure what you mean. We have the following lerp: 

```py
sigmas = get_sigmas(timesteps, n_dim=model_input.ndim, dtype=model_input.dtype)
noisy_model_input = (1.0 - sigmas) * model_input + sigmas * noise
```

You mean it should be:

```py
sigmas = get_sigmas(timesteps, n_dim=model_input.ndim, dtype=model_input.dtype)
noisy_model_input = (1.0 - sigmas) * noise + sigmas * model_input
```

> dynamic shift for inference

I guess we don't have to account for that during fine-tuning?

**Comment by zhuole1025** [comment]
Yes, I mean when sigmas=1.0, noisy_model_input should be model_input. The finetuning schedule should be decided by the user to use uniform, logitnormal, or with shift. (I think this is already supported in diffuers so no need to do this anymore)

**Comment by sayakpaul** [comment]
> Yes, I mean when sigmas=1.0, noisy_model_input should be model_input. The finetuning schedule should be decided by the user to use uniform, logitnormal, or with shift. (I think this is already supported in diffuers so no need to do this anymore)

Thanks @zhuole1025. Could you check the training code and confirm if possible. I don't think we train when shift. But would love to implement that, too. Could you provide some guidance?

**Comment by zhuole1025** [comment]
> > Yes, I mean when sigmas=1.0, noisy_model_input should be model_input. The finetuning schedule should be decided by the user to use uniform, logitnormal, or with shift. (I think this is already supported in diffuers so no need to do this anymore)
> 
> Thanks @zhuole1025. Could you check the training code and confirm if possible. I don't think we train when shift. But would love to implement that, too. Could you provide some guidance?

I took a rough look and find almost everything seems good, except for our lumina  predict data - noise, so the objective should be reversed too.

File: examples/dreambooth/train_dreambooth_lora_lumina2.py [hunk 1]
```diff
@@ -0,0 +1,1543 @@
...
+                # Lumina2 reverses the lerp i.e., sigma of 1.0 should mean `model_input`
+                sigmas = get_sigmas(timesteps, n_dim=model_input.ndim, dtype=model_input.dtype)
+                noisy_model_input = (1.0 - sigmas) * noise + sigmas * model_input
+
+                # Predict the noise residual
+                model_pred = transformer(
+                    hidden_states=noisy_model_input,
+                    encoder_hidden_states=prompt_embeds,
+                    encoder_attention_mask=prompt_attention_mask,
+                    timestep=timesteps,
```
[the remark this answers is not in the document]
Comment (zhuole1025) [reply examples/dreambooth/train_dreambooth_lora_lumina2.py:1374]: yes, the input time is in range [0-1]

[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_lumina2.py:1374]: Hmm, how about (1-t)? Should that be incorporated too now that we're using a reversed lerp?

[the remark this answers is not in the document]
Comment (zhuole1025) [reply examples/dreambooth/train_dreambooth_lora_lumina2.py:1374]: sorry I don't get the point. I think current_timestep = 1 - t / self.scheduler.config.num_train_timesteps  can be used here too

[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_lumina2.py:1374]: Yeah that is what I asked.  Thanks! Just to be sure, this along with the lerp of `noisy_model_input = (1.0 - sigmas) * noise + sigmas * model_input` should be right way, correct?

[the remark this answers is not in the document]
Comment (zhuole1025) [reply examples/dreambooth/train_dreambooth_lora_lumina2.py:1374]: sorry for the confusion. This place is not needed if you already reverse the input timesteps above— either you reverse the lerp or reverse the input timesteps

**Comment by sayakpaul** [comment]
@zhuole1025 thanks for your help and guidance so far. https://wandb.ai/sayakpaul/dreambooth-lumina2-lora/runs/d7bxrrnm is the latest run:

```bash
accelerate launch train_dreambooth_lora_lumina2.py \
  --pretrained_model_name_or_path=Alpha-VLLM/Lumina-Image-2.0  \
  --dataset_name=Norod78/Yarn-art-style --instance_prompt="a puppy, yarn art style" \
  --output_dir=trained-lumina2-lora-yarn \
  --mixed_precision="bf16" \
  --resolution=1024 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=4 \
  --use_8bit_adam \
  --learning_rate=1e-5 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=700 \
  --final_validation_prompt="a puppy in a pond, yarn art style" \
  --seed="0"
```

Do the latest changes look good to you? I think I have resolved all the feedback. And if you do a run on your end on any subject you want, that would be awesome!

File: examples/dreambooth/train_dreambooth_lora_lumina2.py [hunk 2]
```diff
@@ -0,0 +1,1557 @@
...
+                # Convert images to latent space
+                if args.cache_latents:
+                    model_input = latents_cache[step].sample()
+                else:
+                    vae = vae.to(accelerator.device)
+                    pixel_values = batch["pixel_values"].to(dtype=vae.dtype)
+                    model_input = vae.encode(pixel_values).latent_dist.sample()
+                    if args.offload:
+                        vae = vae.to("cpu")
+                model_input = model_input * vae_config_scaling_factor
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_lumina2.py:1353]: Need to account for this too:
https://github.com/huggingface/diffusers/blob/6fe05b9b93593bca41afac79b32b7a23526b0e96/src/diffusers/pipelines/lumina2/pipeline_lumina2.py#L747

So the current code should become:

```py
vae_config_scaling_factor = vae.config.scaling_factor
vae_config_shift_factor = vae.config.shift_factor
model_input = (model_input - vae_config_shift_factor) * vae_config_scaling_factor
```

File: examples/dreambooth/train_dreambooth_lora_lumina2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (a-r-r-o-w) [reply examples/dreambooth/train_dreambooth_lora_lumina2.py:1390]: ✅

**Review by a-r-r-o-w — APPROVED** [review]
LGTM!
